### PR TITLE
[Code health] Perform cppcheck cleanup

### LINKED ISSUE #2297: [CI] Add a C++ static code analyser in the build
This is needed for security audits on the code.

parent issue: #2282

### PR DESCRIPTION
**Pull request by chusitoo**
Fixes #2297

## Changes

- Cleaned up code based on current settings for CI to pass
- Some warnings are either false positives and/or fail in v2.13 but no longer reported in v2.14+, although this would require building cppcheck from source
- Also suppressed _exporters/etw/include/opentelemetry/exporters/etw/TraceLoggingDynamic.h_ as it has dependencies on Windows macros (partially covered by microsoft_sal.cfg but not entirely). An alternative to suppressing it, maybe as a future improvement, could be to supply a custom cfg file with rules for cppcheck to ingest.

## Remaining issues

To summarize the 9 outstanding false positives currently being reported by cppcheck v2.13, can be mitigated by using a newer release, more precisely:

**v2.14** fixes
- ext/include/opentelemetry/ext/http/client/curl/http_client_curl.h:310:8: style: Virtual function 'CancelAllSessions' is called from destructor '~HttpClient()' at line 206. Dynamic binding is not used. [virtualCallInConstructor]
- sdk/include/opentelemetry/sdk/logs/batch_log_record_processor.h:88:8: style: Virtual function 'Shutdown' is called from destructor '~BatchLogRecordProcessor()' at line 325. Dynamic binding is not used. [virtualCallInConstructor]
- sdk/include/opentelemetry/sdk/logs/multi_log_record_processor.h:46:8: style: Virtual function 'ForceFlush' is called from destructor '~MultiLogRecordProcessor()' at line 33. Dynamic binding is not used. [virtualCallInConstructor]
- sdk/include/opentelemetry/sdk/logs/multi_log_record_processor.h:56:8: style: Virtual function 'Shutdown' is called from destructor '~MultiLogRecordProcessor()' at line 34. Dynamic binding is not used. [virtualCallInConstructor]
- sdk/include/opentelemetry/sdk/trace/batch_span_processor.h:86:8: style: Virtual function 'Shutdown' is called from destructor '~BatchSpanProcessor()' at line 322. Dynamic binding is not used. [virtualCallInConstructor]
- sdk/include/opentelemetry/sdk/trace/multi_span_processor.h:124:8: style: Virtual function 'Shutdown' is called from destructor '~MultiSpanProcessor()' at line 140. Dynamic binding is not used. [virtualCallInConstructor]
- sdk/include/opentelemetry/sdk/trace/simple_processor.h:77:8: style: Virtual function 'Shutdown' is called from destructor '~SimpleSpanProcessor()' at line 88. Dynamic binding is not used. [virtualCallInConstructor]

**v2.16** fixes
- sdk/include/opentelemetry/sdk/common/circular_buffer.h:188:16: portability: 'data' is of type 'void *'. When using void pointers in calculations, the behaviour is undefined. [arithOperationsOnVoidPointer]
- sdk/include/opentelemetry/sdk/common/circular_buffer.h:190:50: portability: 'data' is of type 'void *'. When using void pointers in calculations, the behaviour is undefined. [arithOperationsOnVoidPointer]

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [X] Changes in public API reviewed

### CODE REVIEW

File: .github/workflows/cppcheck.yml
Comment (marcalff) [inline .github/workflows/cppcheck.yml:12]: ubuntu-latest is only 22.04 at this point, because 24.04 is not yet flagged stable.

Please try ubuntu-24.04 instead, in hope to get a more recent cppcheck.

File: .github/workflows/cppcheck.yml
Comment (marcalff) [inline .github/workflows/cppcheck.yml:41]: Please add another statement, to run `cppcheck --version`, so we can see the actual version used in CI logs.
This will help debugging.

**Review by marcalff — COMMENTED** [review]
LGTM, thanks for the contribution.

Please add a separate PR for the CI alone, which can be merged right away in the main branch. Comment out the last step, to not fail the build on errors, this will be enforced when cleanup is completed.

This will allow us reviewers to actually see the current cppcheck errors in the log, which will facilitate review, as we won't have to ask "why ?" or "what was the error ?" on various lines fixed.

After CI is merged, this PR will contain only the cleanup fixes then.

Comments on actual code cleanup to follow.

File: api/include/opentelemetry/baggage/baggage_context.h
Comment (marcalff) [inline api/include/opentelemetry/baggage/baggage_context.h:30]: Providing details because this is not trivial, as justification, for all reviewers:

* ABI change (the prototype is different)
* on a helper API
* which is inlined anyway, expanded in the caller compile unit

Hence:

* there is **no** ABI change overall, as no API implemented by the SDK is touched

Approved, this is safe IMO.

File: .github/workflows/cppcheck.yml
Comment (chusitoo) [reply .github/workflows/cppcheck.yml:12]: Definitely! It did not make a difference in terms of issues flagged but the versions of cppcheck were different. 

In ubuntu-latest it installed v2.7 whereas with ubuntu-24.04 we get 2.13, which is what I was already testing with locally.

File: api/include/opentelemetry/context/context.h
Comment (marcalff) [inline api/include/opentelemetry/context/context.h:33]: Approved as ABI safe, implementation change only.

File: api/include/opentelemetry/context/context.h
Comment (marcalff) [inline api/include/opentelemetry/context/context.h:39]: Approved as ABI safe, implementation change only.

File: .github/workflows/cppcheck.yml
Comment (chusitoo) [reply .github/workflows/cppcheck.yml:41]: Done in CI only PR #3151

File: api/include/opentelemetry/context/context.h
Comment (marcalff) [inline api/include/opentelemetry/context/context.h:99]: Why reordering the members here ?

Comment (chusitoo) [reply api/include/opentelemetry/context/context.h:99]: If I recall, it started complaining about member ordering when I added the copy ctor that it asked [here](https://github.com/open-telemetry/opentelemetry-cpp/actions/runs/11902741932/job/33168372488?pr=3151#step:4:55)

I just re-ran the the test with `next_` back up above `key_length_` and cppcheck seems to be more accepting of it, so it must've been something else while the changes were still work in progress on my end.

Do you want me to revert this ordering?

File: api/include/opentelemetry/context/context.h
Comment (marcalff) [inline api/include/opentelemetry/context/context.h:137]: Ok to have a copy constructor, this clears the cppcheck warning, and is safer in case of refactoring.

I suspect this code to be unreachable.

File: api/include/opentelemetry/context/propagation/global_propagator.h
Comment (marcalff) [inline api/include/opentelemetry/context/propagation/global_propagator.h:35]: Helper inlined.
Approved as ABI safe.

File: api/include/opentelemetry/context/runtime_context.h
Comment (marcalff) [inline api/include/opentelemetry/context/runtime_context.h:156]: Helper inlined.
Approved as ABI safe.

File: api/include/opentelemetry/logs/event_id.h
Comment (marcalff) [inline api/include/opentelemetry/logs/event_id.h:23]: ABI safe, internal implementation change.

File: api/include/opentelemetry/logs/provider.h
Comment (marcalff) [inline api/include/opentelemetry/logs/provider.h:42]: Helper inlined.
Approved as ABI safe.

File: api/include/opentelemetry/logs/provider.h
Comment (marcalff) [inline api/include/opentelemetry/logs/provider.h:63]: Helper inlined.
Approved as ABI safe.

File: api/include/opentelemetry/metrics/provider.h
Comment (marcalff) [inline api/include/opentelemetry/metrics/provider.h:41]: Helper inlined.
Approved as ABI safe.

File: api/include/opentelemetry/plugin/tracer.h
Comment (marcalff) [inline api/include/opentelemetry/plugin/tracer.h:23]: Here changing the span constructor does not change the object layout or virtual table, should be ok.

This plugin class in particular is not inherited from in the SDK, changing the constructor is ok.

Approved as ABI safe.

Beside, plugin/tracer.h is not used as far as I know.

File: api/include/opentelemetry/trace/default_span.h
Comment (marcalff) [inline api/include/opentelemetry/trace/default_span.h:66]: Not sure if this one is safe, due to:

```
opentracing-shim/src/tracer_shim.cc:        new opentelemetry::trace::DefaultSpan(span_context)};
```

File: api/include/opentelemetry/trace/context.h
Comment (marcalff) [inline api/include/opentelemetry/trace/context.h:38]: Helper inlined.
Approved as ABI safe.

File: api/include/opentelemetry/trace/span_context.h
Comment (marcalff) [inline api/include/opentelemetry/trace/span_context.h:44]: Class SpanContext is used almost everywhere.

Not sure if this change breaks the ABI or not, this needs more analysis.

File: api/include/opentelemetry/context/context.h
Comment (marcalff) [reply api/include/opentelemetry/context/context.h:99]: Yes, please revert. Better not take chances with binary layout if we can avoid it.

In theory, theory and practice are the same.
In practice, they are different.

Comment (marcalff) [reply api/include/opentelemetry/context/context.h:99]: Thinking more about it, I think changing the order of members is a breaking change:

* an instrumented library A built with otel-cpp v1.17.0 (without this change)
* an instrumented library B built with otel-cpp v1.18.0 (with this change)

should see the same bits in the same place, when manipulating a runtime context using their own copy of `class Context`, which includes `DataList`, when both libraries A and B are linked in the same application.

The runtime context is shared between all instrumented libraries, because it comes from a global singleton.

Same for the baggage propagator.

**Comment by marcalff** [comment]
@chusitoo The CI is now merged to main, please merge main into this PR to resolve the conflicts.

File: api/include/opentelemetry/context/context.h
Comment (chusitoo) [reply api/include/opentelemetry/context/context.h]: I remember now; the issue came up not with cppcheck but with `-Werror=reorder` in maintainer mode CI checks and it felt more natural to initialize the key and key length back-to-back so I had swapped the order of the member variables instead the order of init in the copy ctor.

In any case, they are reverted to their original order and the change is in now in the copy constructor to respect the original order of initialization.

File: api/include/opentelemetry/trace/span_context.h
Comment (chusitoo) [reply api/include/opentelemetry/trace/span_context.h]: I think there would be a way to compromise on the direct requirement from cppcheck in order to ensure ABI compatibility. I reverted this change (pass by const ref -> pass by value) and move the copy of that shared_ptr into the member variable. 

This should be comparable since it doesn't waste making an extra copy and incrementing/decrementing the use count. 
Verified that cppcheck agrees with this change and no longer reports it as problematic.

File: api/include/opentelemetry/trace/default_span.h
Comment (chusitoo) [reply api/include/opentelemetry/trace/default_span.h]: Reverted this change, just to be on the safe side. Instead, we move the copy into the member variable, similar to the latest change for [span_context.h](https://github.com/open-telemetry/opentelemetry-cpp/pull/3150#discussion_r1849343531)

File: .github/workflows/cppcheck.yml
Comment (chusitoo) [inline .github/workflows/cppcheck.yml]: Should this check be re-enabled now?

File: exporters/otlp/src/otlp_http_client.cc
Comment (marcalff) [inline exporters/otlp/src/otlp_http_client.cc:904]: Here and in other places, the call to result_callback() has moved a bit.

Could you provide more details ?

File: ext/include/opentelemetry/ext/http/client/curl/http_client_curl.h
Comment (marcalff) [inline ext/include/opentelemetry/ext/http/client/curl/http_client_curl.h:310]: Please remove the suppression.

Instead, if this still raises an warning in cppcheck, we need to find where from, and fix the caller code.

File: sdk/include/opentelemetry/sdk/logs/batch_log_record_processor.h
Comment (marcalff) [inline sdk/include/opentelemetry/sdk/logs/batch_log_record_processor.h:88]: Remove the suppression, and identify/fix callers.

File: sdk/include/opentelemetry/sdk/logs/multi_log_record_processor.h
Comment (marcalff) [inline sdk/include/opentelemetry/sdk/logs/multi_log_record_processor.h:57]: Remove the suppression, and identify/fix callers.

File: sdk/include/opentelemetry/sdk/trace/batch_span_processor.h
Comment (marcalff) [inline sdk/include/opentelemetry/sdk/trace/batch_span_processor.h:86]: Remove the suppression, and identify/fix callers.

File: sdk/include/opentelemetry/sdk/trace/multi_span_processor.h
Comment (marcalff) [inline sdk/include/opentelemetry/sdk/trace/multi_span_processor.h:124]: Remove the suppression, and identify/fix callers.

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h
Comment (marcalff) [inline sdk/include/opentelemetry/sdk/trace/simple_processor.h:77]: Remove the suppression, and identify/fix callers.

File: api/include/opentelemetry/trace/default_span.h
Comment (marcalff) [reply api/include/opentelemetry/trace/default_span.h]: Thanks, the change is much better now.

File: .github/workflows/cppcheck.yml
Comment (marcalff) [reply .github/workflows/cppcheck.yml]: Yes, please do.

If there are unresolved warnings left, it is ok to test with some margin, say  `count > 5`.

The point is to prevent new errors to come in with more merges, while giving us time to resolve possibly more complex issues in follow up PRs and ultimately have `count > 0`.

**Review by marcalff — COMMENTED** [review]
Thanks for the cleanup.

All the code under api/ looks good to me, but I would like to have a second opinion (@lalitb , @ThomsonTan , @esigo) due to the risk of breaking the ABI.

See new comments for non api code.

File: exporters/otlp/src/otlp_http_client.cc
Comment (chusitoo) [reply exporters/otlp/src/otlp_http_client.cc:904]: Absolutely. The scan reports an _[accessMoved]_ warning [here](https://github.com/open-telemetry/opentelemetry-cpp/actions/runs/11948772032/job/33306927417#step:4:440)

This is due to the move that happens on higher `OtlpHttpClient::Export()` [L753](https://github.com/open-telemetry/opentelemetry-cpp/pull/3150/files/e9b225295e05a2baf71708e6486036042ed69542#diff-eb7f99ebc152d9cc0f4781cc0a0388907e4d4b58271dba85eba99fd16b880674R753) which looks like it is likely undefined behavior calling it from that point on.

Since the **_moved from_** result_callback appears to be used only to consume the result from `createSession()`, my solution was to move the invocation closer to the call sites where it's been **_moved to_**.

File: ext/include/opentelemetry/ext/http/client/curl/http_client_curl.h
Comment (chusitoo) [reply ext/include/opentelemetry/ext/http/client/curl/http_client_curl.h:310]: Please note that this and all other suppressions of `[virtualCallInConstructor]` are false positives no longer reported in v2.14+, as highlighted in the [changelog summary](https://github.com/danmar/cppcheck/releases/tag/2.14.0).

I will revert, as requested. As a future improvement, using a more updated version of cppcheck (built from source) should mitigated these warnings.

Comment (chusitoo) [reply ext/include/opentelemetry/ext/http/client/curl/http_client_curl.h:310]: @marcalff I am noticing that the changes I had put in circular_buffer.h to suppress both `[arithOperationsOnVoidPointer]` are no longer reported in v2.16, so I am wondering if these false positives waiting to be addressed should be documented somewhere. I think this would be beneficial to avoid repeat investigation when the fix is updating the scanner version.

File: .github/workflows/cppcheck.yml
Comment (chusitoo) [reply .github/workflows/cppcheck.yml]: Made the `WARNING_LIMIT = 10` to match your PR for iwyu part 5 and to accommodate the 9 false positives that are no longer suppressed. I also thought it would be a good idea to make it stand out a bit more when the count is below threshold but not zero, by emitting a warning annotation.

image

Comment (marcalff) [reply .github/workflows/cppcheck.yml]: Very nice.

I am applying this for include-what-you-use as well.

File: exporters/otlp/src/otlp_http_client.cc
Comment (marcalff) [reply exporters/otlp/src/otlp_http_client.cc:904]: Good call on `result_callback`, thanks for finding and fixing this.

**Review by marcalff — APPROVED** [review]
LGTM, excellent cleanup.

@lalitb , @ThomsonTan , @esigo,

Please double check the code under `api/` in particular, due to the risk of breaking the ABI.

I left comments to indicate why I think each change is safe.

File: ext/include/opentelemetry/ext/http/client/curl/http_client_curl.h
Comment (chusitoo) [reply ext/include/opentelemetry/ext/http/client/curl/http_client_curl.h]: Good suggestion, thanks!

To summarize the 9 outstanding false positives currently being reported by cppcheck v2.13, can be mitigated by using a newer release, more precisely:

**v2.14** fixes
- ext/include/opentelemetry/ext/http/client/curl/http_client_curl.h:310:8: style: Virtual function 'CancelAllSessions' is called from destructor '~HttpClient()' at line 206. Dynamic binding is not used. [virtualCallInConstructor]
- sdk/include/opentelemetry/sdk/logs/batch_log_record_processor.h:88:8: style: Virtual function 'Shutdown' is called from destructor '~BatchLogRecordProcessor()' at line 325. Dynamic binding is not used. [virtualCallInConstructor]
- sdk/include/opentelemetry/sdk/logs/multi_log_record_processor.h:46:8: style: Virtual function 'ForceFlush' is called from destructor '~MultiLogRecordProcessor()' at line 33. Dynamic binding is not used. [virtualCallInConstructor]
- sdk/include/opentelemetry/sdk/logs/multi_log_record_processor.h:56:8: style: Virtual function 'Shutdown' is called from destructor '~MultiLogRecordProcessor()' at line 34. Dynamic binding is not used. [virtualCallInConstructor]
- sdk/include/opentelemetry/sdk/trace/batch_span_processor.h:86:8: style: Virtual function 'Shutdown' is called from destructor '~BatchSpanProcessor()' at line 322. Dynamic binding is not used. [virtualCallInConstructor]
- sdk/include/opentelemetry/sdk/trace/multi_span_processor.h:124:8: style: Virtual function 'Shutdown' is called from destructor '~MultiSpanProcessor()' at line 140. Dynamic binding is not used. [virtualCallInConstructor]
- sdk/include/opentelemetry/sdk/trace/simple_processor.h:77:8: style: Virtual function 'Shutdown' is called from destructor '~SimpleSpanProcessor()' at line 88. Dynamic binding is not used. [virtualCallInConstructor]

**v2.16** fixes
- sdk/include/opentelemetry/sdk/common/circular_buffer.h:188:16: portability: 'data' is of type 'void *'. When using void pointers in calculations, the behaviour is undefined. [arithOperationsOnVoidPointer]
- sdk/include/opentelemetry/sdk/common/circular_buffer.h:190:50: portability: 'data' is of type 'void *'. When using void pointers in calculations, the behaviour is undefined. [arithOperationsOnVoidPointer]

File: api/include/opentelemetry/context/context.h
Comment (sjinks) [inline api/include/opentelemetry/context/context.h:143]:
```suggestion
      std::memcpy(key_, other.key_, other.key_length_ * sizeof(char));
```

because we `#include <cstring>`, not `<string.h>`

**Review by lalitb — APPROVED** [review]
reviewed for ABI breaking changes in API. Looks good from that prospective.

**Pull request merged**
